src/native/libs/System.Native/pal_runtimeinformation.c

```diff
@@ -48,8 +53,93 @@ int32_t SystemNative_GetUnixVersion(char* version, int* capacity)
     return 0;
 }
 
-/* Returns an int representing the OS Architecture. -1 if same as process architecture. */
+// Keep in sync with System.Runtime.InteropServices.Architecture enum
+enum
+{
+    ARCH_X86,
+    ARCH_X64,
+    ARCH_ARM,
+    ARCH_ARM64,
+    ARCH_WASM,
+    ARCH_S390X,
+    ARCH_LOONGARCH64,
+    ARCH_ARMV6,
+};
+
 int32_t SystemNative_GetOSArchitecture()
 {
-    return -1;
+#ifdef TARGET_WASM
+    return ARCH_WASM;
+#else
+    int32_t result = -1;
+#ifdef TARGET_SUNOS
+    // On illumos/Solaris, the recommended way to obtain machine
+    // architecture is using `sysinfo` rather than `utsname.machine`.
+
+    char isa[32];
+    if (sysinfo(SI_ARCHITECTURE_K, isa, sizeof(isa)) > -1)
+    {
+        {
+#else
+    {
+        struct utsname _utsname;
+        if (uname(&_utsname) > -1)
+        {
+            char* isa = _utsname.machine;
+#endif
+            // aarch64 or arm64: arm64
+            if (strcmp("aarch64", isa) == 0 || strcmp("arm64", isa) == 0)
+            {
+#ifdef TARGET_OSX
+                int is_translated_process = 0;
+                size_t size = sizeof(is_translated_process);
+                if (sysctlbyname("sysctl.proc_translated", &is_translated_process, &size, NULL, 0) == 0)
+                    result = ARCH_X64;
```

Comment:
Have you been able to validate this on Apple M1?

The host OS architecture for the Apple M1 emulated case is arm64 and the process architecture is x64. This logic looks the other way around.